src/mlpack/core/optimizers/adam/adamirror_update.hpp

```diff
@@ -0,0 +1,144 @@
+ * @author Moksh Jain
+ *
+ * Adamirror optimizer. Optimistic Adam is an an algorithm which uses
+ * Optimistic Mirror Descent with the Adam optimizer.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_ADAM_ADAMIRROR_UPDATE_HPP
+#define MLPACK_CORE_OPTIMIZERS_ADAM_ADAMIRROR_UPDATE_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace optimization {
+
+/**
+ * Adamirror is an optimizer which uses Optmistic Mirror Descent with
+ * the Adam Optimizer.
+ *
+ * For more information, see the following.
+ *
+ * @code
+ * @article{
+ *   author  = {Constantinos Daskalakis, Andrew Ilyas, Vasilis Syrgkanis, 
+ *              Haoyang Zeng},
+ *   title   = {Training GANs with Optimism},
+ *   year    = {2017},
+ *   url     = {https://openreview.net/forum?id=SJJySbbAZ}
+ * }
+ * @endcode
+ */
+class AdamirrorUpdate
+{
+ public:
+  /**
+   * Construct the Adamirror update policy with the given parameters.
+   *
+   * @param epsilon The epsilon value used to initialise the squared gradient
+   *        parameter.
+   * @param beta1 The smoothing parameter.
+   * @param beta2 The second moment coefficient.
+   */
+  AdamirrorUpdate(const double epsilon = 1e-8,
+             const double beta1 = 0.9,
+             const double beta2 = 0.999) :
+    epsilon(epsilon),
+    beta1(beta1),
+    beta2(beta2),
+    iteration(0)
+  {
+    // Nothing to do.
+  }
+
+  /**
+   * The Initialize method is called by SGD Optimizer method before the start of
+   * the iteration update process.
+   *
+   * @param rows Number of rows in the gradient matrix.
+   * @param cols Number of columns in the gradient matrix.
+   */
+  void Initialize(const size_t rows, const size_t cols)
+  {
+    m = arma::zeros<arma::mat>(rows, cols);
+    v = arma::zeros<arma::mat>(rows, cols);
+    g = arma::zeros<arma::mat>(rows, cols);
+  }
+
+  /**
+   * Update step for Adamirror.
+   *
+   * @param iterate Parameters that minimize the function.
+   * @param stepSize Step size to be used for the given iteration.
+   * @param gradient The gradient matrix.
+   */
+  void Update(arma::mat& iterate,
+              const double stepSize,
+              const arma::mat& gradient)
+  {
+    // Increment the iteration counter variable.
+    ++iteration;
+
+    // And update the iterate.
+    m *= beta1;
+    m += (1 - beta1) * gradient;
+
+    v *= beta2;
+    v += (1 - beta2) * arma::square(gradient);
+
+    arma::mat mCorrected = m / (1.0 - std::pow(beta1, iteration));
+    arma::mat vCorrected = v / (1.0 - std::pow(beta2, iteration));
+
+    arma::mat update = mCorrected / (arma::sqrt(vCorrected) + epsilon);
+
+    iterate -= (2 * stepSize * update - stepSize * g);
+
+    g = update;
```

Comment:
We might be able to avoid an extra copy here, e.g. by using ``g = std::move(update)``.

Reply:
I did not know that `std::move` avoids the extra copy. Thanks. I'll change it. :)